### PR TITLE
Fix four issues in archive.extracted state

### LINKED ISSUE #39868: archive.extracted issue when source_hash_update=True and extracted files does not exist
### Description of Issue/Question
Not sure if I'm not using archive.extracted properly or if this is a bug, I think that it should check for the tarball content besides the hashes are not changed and extract if files are missing.

### Setup
```sls
perlbrew:
  archive.extracted:
    - name: /
    - source: http://perl.company.com/perlbrew.tar.gz
    - source_hash: http://perl.company.com/sha256sum.txt
    - source_hash_update: True
    - overwrite: True
```

### Output
```
[DEBUG   ] LazyLoaded archive.extracted
[INFO    ] Running state [/] at time 04:15:28.018056
[INFO    ] Executing state archive.extracted for /
[DEBUG   ] Requesting URL http://perl.company.com/sha256sum.txt using GET method
[DEBUG   ] file.extract_hash: Extracting any supported hash for file matching: http://perl.company.com/perlbrew.tar.gz
[DEBUG   ] file.extract_hash: Line '009f621fbadbea9999c4a33fcb3c658703315344a4e9c0be4e3685a94d262d73  perlbrew.tar.gz' matches source 'http://perl.company.com/perlbrew.tar.gz'
[DEBUG   ] file.extract_hash: Line '44e5284d273dc486ec40ab3360a8bfcee6b83a06529ddfb1202dbdf905c39665  company_mfa.tar.gz' contains sha256 hash '44e5284d273dc486ec40ab3360a8bfcee6b83a06529ddfb1202dbdf905c39665', but line did not meet the search criteria
[DEBUG   ] file.extract_hash: Line '951c397e88fec12933657751b7bda8b21a4b683a3591cd12c8fb84f54ecacf47  company_transtool_api.tar.gz' contains sha256 hash '951c397e88fec12933657751b7bda8b21a4b683a3591cd12c8fb84f54ecacf47', but line did not meet the search criteria
[DEBUG   ] file.extract_hash: Returning sha256 hash '009f621fbadbea9999c4a33fcb3c658703315344a4e9c0be4e3685a94d262d73' as a match of http://perl.company.com/perlbrew.tar.gz
[INFO    ] Hash 009f621fbadbea9999c4a33fcb3c658703315344a4e9c0be4e3685a94d262d73 has not changed
[INFO    ] Completed state [/] at time 04:15:28.138946 duration_in_ms=120.89
```
### Steps to Reproduce Issue
Run the above state, then remove the extracted tarball path and run the state again. Salt will compare hashes and not extract as stated in documentation:

> Salt now uses a function called archive.list to get a list of files/directories in the archive. Using this information, the state can now check the minion to see if any paths are missing, and know whether or not the archive needs to be extracted. This makes the if_missing argument unnecessary in most use cases.

### Versions Report
```
Salt Version:
           Salt: 2016.11.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.13 (default, Jan 12 2017, 01:15:13)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 16.0.1
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.1.5
 
System Versions:
           dist:   
        machine: amd64
        release: 10.3-RELEASE-p11
         system: FreeBSD
        version: Not Installed
```

### PR DESCRIPTION
**Pull request by terminalmage**
1. The trim_output argument was ignored for archives extracted using tarfile.

2. ``file.get_source_sum`` would fail if ``source`` is a list.

3. The checksum was being updated before we checked to see if it matched, effectively keeping us from detecting changes to the hash.

4. When ``source_hash_update`` is ``True``, and the archive is extracted, and files are later removed from the extraction dir, the state would not re-extract the archive unless the ``source_hash`` had changed (#39868).

Resolves #39868.

### CODE REVIEW

File: salt/states/archive.py
Comment (isbm) [inline salt/states/archive.py]: @terminalmage I'm just thinking maybe it is easier to comprehend by first discarding `bool`?
```python
trim_output = 100 if type(trim_output) == bool and trim_output else False
if trim_output and type(trim_output) != bool:
    try:
        trim_output = int(trim_output)
    except TypeError:
        ...
```

**Review by isbm — APPROVED** [review]
Seems OK (I wrote one comment first, but then deleted, since I was wrong). You love newlines and `\` though... :-)

**Comment by terminalmage** [comment]
What can I say, I like readability :man_shrugging:

**Comment by cachedout** [comment]
@terminalmage Still pending changes here?

**Comment by terminalmage** [comment]
@cachedout yes, I'm working with a community member to test updates to this state, the conversation is not taking place in this PR comment thread. I'll update the label when it's ready to merge.

**Comment by terminalmage** [comment]
I've pushed some changes that should have this ready for merging, but I'm waiting for testing/signoff from @jerrykan (see https://github.com/saltstack/salt/issues/40484#issuecomment-292661994).

**Comment by terminalmage** [comment]
Some failed unit tests due to the introduction of the ``hash_type`` option in the code (needed to be mocked in the tests).

**Comment by terminalmage** [comment]
Should be fixed with my latest push.

**Comment by Ch3LL** [comment]
@terminalmage is this ready for review and merge or are we still waiting on user?

**Comment by terminalmage** [comment]
See #40484 for conversation with the community member I've been working with. It looks like I have one lingering issue to resolve in this PR before it can be merged.

**Comment by terminalmage** [comment]
OK, this should be good to merge pending test results.

**Pull request merged**
